feat: Implement RLP Encoding and Decoding

Quentash opened the pull request:
Implementation of RLP encoding and decoding based on https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/ doc.

<!--- Please provide a general summary of your changes in the title above -->

## Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #212 

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this does introduce a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

<!-- Any other information that is important to this PR, such as screenshots of how the component looks before and after the change. -->

LucasLvy commented on src/encoding/src/rlp.cairo:
```diff
@@ -0,0 +1,183 @@
+    /// # Errors
+    /// * empty input - if the input is empty
+    /// * input too short - if the input is too short for a given
+    fn decode_type(input: Span<u8>) -> (RLPType, u32, u32) {
```
I think this should return a result instead of panics

Quentash replied:
Ho ok ! Will do !

LucasLvy reviewed (changes requested):
Same for other functiosn

Quentash commented:
Created a errors.cairo file inside encoding crate.
Changed panics to Results.

LucasLvy commented on src/encoding/src/errors.cairo:
```diff
@@ -0,0 +1,18 @@
+// LENGTH
+const RLP_EMPTY_INPUT: felt252 = 'KKT: EmptyInput';
+const RLP_INPUT_TOO_SHORT: felt252 = 'KKT: InputTooShort';
```
Don't prefix that with KKT

LucasLvy commented on src/encoding/src/errors.cairo:
```diff
@@ -0,0 +1,18 @@
+
+#[derive(Drop, Copy, PartialEq)]
+enum RLPError {
+    EmptyInput: felt252,
+    InputTooShort: felt252,
```
The names are descriptive enough so you don't need to add any data here

LucasLvy commented on src/encoding/src/errors.cairo:
```diff
@@ -0,0 +1,18 @@
+    InputTooShort: felt252,
+}
+
+impl RLPErrorIntoU256 of Into<RLPError, u256> {
+    fn into(self: RLPError) -> u256 {
+        match self {
+            RLPError::EmptyInput(error_message) => error_message.into(),
+            RLPError::InputTooShort(error_message) => error_message.into(),
+        }
+    }
+}
```
why do you need that ?

Quentash replied:
I don't, sorry ! It is removed now

LucasLvy commented on src/encoding/src/rlp.cairo:
```diff
@@ -0,0 +1,184 @@
+        }
+
+        let mut output: Array<u8> = Default::default();
+        let item = input.pop_front().unwrap();
```
Add a comment saying why it's safe to unwrap

LucasLvy commented on src/encoding/src/rlp.cairo:
```diff
@@ -0,0 +1,184 @@
+                        output.append(0xf7 + len_in_bytes.len().try_into().unwrap());
+                        output.concat_span(len_in_bytes);
+                    } else {
+                        output.append(0xc0 + payload_len.try_into().unwrap());
```
same here

LucasLvy commented on src/encoding/src/rlp.cairo:
```diff
@@ -0,0 +1,184 @@
+            return Result::Ok(input);
+        } else if len < 56 {
+            let mut encoding: Array<u8> = Default::default();
+            encoding.append(0x80 + len.try_into().unwrap());
```
same here

LucasLvy commented on src/encoding/src/rlp.cairo:
```diff
@@ -0,0 +1,184 @@
+            let mut encoding: Array<u8> = Default::default();
+            let len_as_bytes = len.to_bytes();
+            let len_bytes_count = len_as_bytes.len();
+            let prefix = 0xb7 + len_bytes_count.try_into().unwrap();
```
same here

LucasLvy commented on src/numeric/src/integers.cairo:
```diff
@@ -0,0 +1,77 @@
+use alexandria_math::BitShift;
+
+#[generate_trait]
+impl U32Impl of U32Trait {
```
probably worth converting that trait in a generic trait that can be implemented over any type

Quentash commented:
- Simplified errors handling and therefore deleted errors.cairo.
- Moved RLPError into rlp.cairo
- Made a generic trait but not sure about the name, do you have any suggestion ?
- Added tests for the U32Impl in integers.cairo
- Corrected list encoding tests variables naming
- Added comments to explain the safe unwraps.

LucasLvy reviewed (changes requested):
remove the unnecessary `(())` and either remove the unwraps or say why they're safe plz

Quentash commented on src/encoding/src/rlp.cairo:
```diff
@@ -0,0 +1,197 @@
+                return Result::Err(RLPError::InputTooShort(()));
+            }
+            let string_len_bytes = input.slice(1, len_bytes_count);
+            let string_len: u32 = UIntBytes::from_bytes(string_len_bytes).unwrap();
```
My very bad, I missed those !!

Quentash replied:
Having to justify the safety of those unwrap made me realize they weren't actually safe.
The problem being that the payload length can be on 8 Bytes ( therefore an u64 ). But arrays don't seem to be able to hold more than MAXU32 values. ( I am apparently unable to test this with my computers :s )

I therefore added a check on the amount of bytes representing the payload length and currently return an error if the payload is > 4 bytes.
But I am not convinced by this because it doesn't allow us to decode very long payloads. I ignore how frequently we are supposed to decode payload with a length above MAXU32 in real life but still, it isn't compliant with the official RLP doc and I don't know how to make it compliant right now using vanilla arrays/spans.

I hope I'm being clear, if I sound confusing, feel free to reach me on TG/Disc/whatever, I'd be very glad to explain the problem !

LucasLvy replied:
you can use a 2d array

LucasLvy replied:
but maybe this should be in another pr idk wdyt ?

Quentash commented:
- Modified error handling for PayloadTooLong using ok_or() instead of the unsafe unwrap()

LucasLvy commented on src/encoding/src/tests/rlp_test.cairo:
```diff
@@ -0,0 +1,3285 @@
+    let res = RLPTrait::decode(ArrayTrait::new().span());
+
+    assert(res.is_err(), 'Should have failed');
+    assert(res.unwrap_err() == RLPError::EmptyInput(()), 'err != EmptyInput');
```
same for all enum variants plz
```suggestion
    assert(res.unwrap_err() == RLPError::EmptyInput, 'err != EmptyInput');
```

LucasLvy commented on src/encoding/src/rlp.cairo:
```diff
@@ -0,0 +1,198 @@
+            return Result::Ok(encoding.span());
+        } else {
+            let mut encoding: Array<u8> = Default::default();
+            // The payload length being a u32, the length in bytes
+            // will maximum be equal to 4, making the unwrap safe
```
this should go 2 lines below

LucasLvy commented on src/encoding/src/tests/rlp_test.cairo:
```diff
@@ -0,0 +1,3285 @@
+    let mut i = 0;
+    loop {
+        if i == 0x80 {
+            break ();
```
```suggestion
            break;
```

LucasLvy commented on src/numeric/src/integers.cairo:
```diff
@@ -0,0 +1,82 @@
+        let mut i: u32 = 0;
+        loop {
+            if i == len {
+                break ();
```
same for all the breaks
```suggestion
                break;
```

Quentash commented:
- Moved comments right above their appropriate unwrap()
- Removed once and for all the (()) ( my very bad )
